node/src/components/storage/lmdb_ext.rs

```diff
@@ -0,0 +1,157 @@
+///
+/// Note that accessing a storage with an incompatible version of this software is also considered a
+/// case of corruption.
+#[derive(Debug, Error)]
+pub enum LmdbExtError {
+    /// The internal database is corrupted and can probably not be salvaged.
+    #[error("internal storage corrupted: {0}")]
+    LmdbCorrupted(lmdb::Error),
+    /// The data stored inside the internal database is corrupted or formatted wrong.
+    #[error("internal data corrupted: {0}")]
+    DataCorrupted(Box<dyn std::error::Error + Send + Sync>),
+    /// A resource has been exhausted a runtime, restarting (potentially with different settings)
+    /// might fix the problem. Storage integrity is still intact.
+    #[error("storage exhausted resource (but still intact): {0}")]
+    ResourceExhausted(lmdb::Error),
+    /// Error neither corruption nor resource exhaustion occured, likely a programming error.
+    #[error("unknown LMDB or serialization error, likely from a bug: {0}")]
+    Other(Box<dyn std::error::Error + Send + Sync>),
+}
+
+// Classifies an `lmdb::Error` according to our scheme. This one of the rare cases where we accept a
+// blanked `From<>` implementation for error type conversion.
+impl From<lmdb::Error> for LmdbExtError {
+    fn from(lmdb_error: lmdb::Error) -> Self {
+        match lmdb_error {
+            lmdb::Error::PageNotFound
+            | lmdb::Error::Corrupted
+            | lmdb::Error::Panic
+            | lmdb::Error::VersionMismatch
+            | lmdb::Error::Invalid
+            | lmdb::Error::Incompatible => LmdbExtError::LmdbCorrupted(lmdb_error),
+
+            lmdb::Error::MapFull
+            | lmdb::Error::DbsFull
+            | lmdb::Error::ReadersFull
+            | lmdb::Error::TlsFull
+            | lmdb::Error::TxnFull
+            | lmdb::Error::CursorFull
+            | lmdb::Error::PageFull
+            | lmdb::Error::MapResized => LmdbExtError::ResourceExhausted(lmdb_error),
+
+            lmdb::Error::NotFound
+            | lmdb::Error::BadRslot
+            | lmdb::Error::BadTxn
+            | lmdb::Error::BadValSize
+            | lmdb::Error::BadDbi
+            | lmdb::Error::KeyExist
+            | lmdb::Error::Other(_) => LmdbExtError::Other(Box::new(lmdb_error)),
+        }
+    }
+}
+
+/// Additional methods on transaction.
+pub(super) trait TransactionExt {
+    /// Helper function to load a value from a database.
+    fn get_value<K: AsRef<[u8]>, V: DeserializeOwned>(
+        &mut self,
+        db: Database,
+        key: &K,
+    ) -> Result<Option<V>, LmdbExtError>;
+}
+
+/// Additional methods on write transactions.
+pub(super) trait WriteTransactionExt {
+    /// Helper function to write a value to a database.
+    ///
+    /// Returns `true` if the value has actually been written, `false` if the key already existed.
+    ///
+    /// Setting `overwrite` to true will cause the value to always be written instead.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a database error occurs.
```

Comment:
When would serialization legit fail? We are only serializing our own types, which should always serialize.

From a different perspective, what can the user of the component reasonably do to avoid or handle this?

Reply:
I am also removing the `expect` there, I guess. Let's propagate the serialization problem upwards, because we can. It'll be an `Other` error.

Reply:
So, there's no more panics now :)